### PR TITLE
Sanitize cypher errors from neo4j driver

### LINKED ISSUE #278: Neo4j Error Records Contain Sensitive Data
**Description:**

BloodHound redacts sensitive information from logged cypher queries by replacing them with ``$STRIPPED`` in the logstrings, e.g.:

```
bloodhound-bloodhound-1  | {"level":"info","query":"match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[r:MemberOf*1..]->(g:Group) with p1 match p2 = (u2)-[r:MemberOf*1..]->(g:Group) return p1, p2","time":"2023-12-19T12:02:39.893419034Z","message":"Executing user cypher query"}
```

If a neo4j driver error occurs, the originally stripped information is not replaced in the error record of the neo4j driver (redaction done by me):

```
bloodhound-bloodhound-1  | {"level":"info","query":"match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[r:MemberOf*1..]->(g:Group) with p1 match p2 = (u2)-[r:MemberOf*1..]->(g:Group) return p1, p2","time":"2023-12-19T12:02:39.893419034Z","message":"Executing user cypher query"}
bloodhound-bloodhound-1  | {"level":"info","time":"2023-12-19T12:02:39.893514766Z","message":"Cypher query cost is: 21.00. Reduction factor for query is: 4. Available timeout for query is now set to: 7.50 seconds"}
bloodhound-bloodhound-1  | {"level":"warn","time":"2023-12-19T12:02:47.787096395Z","message":"Writing API Error. Status: 500. Message: [{ driver error: Neo4jError: Neo.ClientError.Transaction.TransactionTimedOut (The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. ) - query: match (u1:User {domain: \"<DOMAIN1>\"}), (u2:User {domain: \"<DOMAIN2>\"}) where u1.samaccountname <> \"krbtgt\" and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[r:MemberOf*1..]->(g:Group) with p1 match p2 = (u2)-[r:MemberOf*1..]->(g:Group) return p1, p2}]"}
```

**Component(s) Affected:**

- [ ] UI
- [ ] API
- [ ] Neo4j
- [ ] PostgreSQL
- [ ] Data Collector (SharpHound, AzureHound)
- [X] Other (tooling, documentation, etc.)

**Steps to Reproduce:**

1. Create a cypher query that produces a neo4j error, e.g. by creating a query with higher database cost. Queries with a high db cost have a limited timeout, if the query cannot be completed the error message will contain the domain names.

**Expected Behavior:**

Sensitive information like domain names should be redacted in all error records.

**Actual Behavior:**

Sensitive information is not redacted in error records of the neo4j driver.

**Screenshots/Code Snippets/Sample Files:**

```
bloodhound-bloodhound-1  | {"level":"info","query":"match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[r:MemberOf*1..]->(g:Group) with p1 match p2 = (u2)-[r:MemberOf*1..]->(g:Group) return p1, p2","time":"2023-12-19T12:02:39.893419034Z","message":"Executing user cypher query"}
bloodhound-bloodhound-1  | {"level":"info","time":"2023-12-19T12:02:39.893514766Z","message":"Cypher query cost is: 21.00. Reduction factor for query is: 4. Available timeout for query is now set to: 7.50 seconds"}
bloodhound-bloodhound-1  | {"level":"warn","time":"2023-12-19T12:02:47.787096395Z","message":"Writing API Error. Status: 500. Message: [{ driver error: Neo4jError: Neo.ClientError.Transaction.TransactionTimedOut (The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. ) - query: match (u1:User {domain: \"<DOMAIN1>\"}), (u2:User {domain: \"<DOMAIN2>\"}) where u1.samaccountname <> \"krbtgt\" and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[r:MemberOf*1..]->(g:Group) with p1 match p2 = (u2)-[r:MemberOf*1..]->(g:Group) return p1, p2}]"}
```

**Environment Information:**

BloodHound: latest

**Potential Solution (Optional):**

Error records need to be parsed for sensitive information before writing them to log or stdout.

**Contributor Checklist:**

- [X] I have searched the issue tracker to ensure this bug hasn't been reported before or is not already being addressed.
- [X] I have provided clear steps to reproduce the issue.
- [X] I have included relevant environment information details.
- [ ] I have attached necessary supporting documents.
- [ ] I have checked that any JSON files I am attempting to upload to BloodHound are valid.

### PR DESCRIPTION
**Pull request by juggernot325**
## Description

This PR adds sanitization to errors returned from the neo4j driver for `dawgs`. This fixes #278.

I also noticed that the `IsNeoTimeoutError` was incorrectly returning false when the time out came from the neo4j driver from `dawgs`. I updated that function to properly handle that case and additional test cases, as well.

## Motivation and Context

* Prevent leaking potentially sensitive information in logs

## How Has This Been Tested?

* I reduced the timeout for cypher queries to a low value and ran a complex query to verify sensitive values were stripped
* Added additional test cases for `IsNeoTimeoutError`

## Output:

Before:

```
bhe-api-1       | {"level":"info","user_id":"7f63d6b4-c472-43f2-9ed1-5fac1a0fa6e8","remote_addr":"172.30.0.5:56408","proto":"HTTP/1.1","referer":"http://bhe.localhost/ui/mockServiceWorker.js","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36","request_id":"42f302d7-b728-4643-8496-6fcc8e606a77","request_bytes":0,"response_bytes":228,"status":200,"elapsed":15.167292,"time":"2024-03-26T18:00:48.440332673Z","message":"GET /api/v2/available-domains"}
bhe-api-1       | {"level":"info","query":"match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[*1..]->(g:Group) with p1 match p2 = (u2)-[*1..]->(g:Group) return p1, p2","time":"2024-03-26T18:00:49.708330924Z","message":"Executing user cypher query"}
bhe-api-1       | {"level":"info","time":"2024-03-26T18:00:49.708374591Z","message":"Cypher query cost is: 25.00. Reduction factor for query is: 5. Available timeout for query is now set to: 0.20 seconds"}
bhe-api-1       | {"level":"warn","time":"2024-03-26T18:00:50.088584382Z","message":"Writing API Error. Status: 500. Message: [{ driver error: Neo4jError: Neo.ClientError.Transaction.TransactionTimedOut (The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. ) - query: match (u1:User {domain: \"ESC6.LOCAL\"}), (u2:User {domain: \"ESC3.LOCAL\"}) where u1.samaccountname <> \"krbtgt\" and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[*1..]->(g:Group) with p1 match p2 = (u2)-[*1..]->(g:Group) return p1, p2}]"}
```

After:
```
bhe-api-1       | {"level":"info","user_id":"7f63d6b4-c472-43f2-9ed1-5fac1a0fa6e8","remote_addr":"172.31.0.4:58282","proto":"HTTP/1.1","referer":"http://bhe.localhost/ui/mockServiceWorker.js","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36","request_id":"bf768d22-5a8e-4c83-b971-bb1902746ab6","request_bytes":0,"response_bytes":228,"status":200,"elapsed":14.926209,"time":"2024-03-26T18:16:18.675574673Z","message":"GET /api/v2/available-domains"}
bhe-api-1       | {"level":"info","query":"match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[*1..]->(g:Group) with p1 match p2 = (u2)-[*1..]->(g:Group) return p1, p2","time":"2024-03-26T18:16:20.867006383Z","message":"Executing user cypher query"}
bhe-api-1       | {"level":"info","time":"2024-03-26T18:16:20.867053424Z","message":"Cypher query cost is: 25.00. Reduction factor for query is: 5. Available timeout for query is now set to: 0.20 seconds"}
bhe-api-1       | {"level":"warn","time":"2024-03-26T18:16:22.205627425Z","message":"Writing API Error. Status: 500. Message: [{ driver error: Neo4jError: Neo.ClientError.Transaction.TransactionTimedOut (The transaction has been terminated. Retry your operation in a new transaction, and you should see a successful result. The transaction has not completed within the specified timeout (dbms.transaction.timeout). You may want to retry with a longer timeout. ) - query: match (u1:User {domain: $STRIPPED}), (u2:User {domain: $STRIPPED}) where u1.samaccountname <> $STRIPPED and u1.samaccountname = u2.samaccountname with u2 match p1 = (u2)-[*1..]->(g:Group) with p1 match p2 = (u2)-[*1..]->(g:Group) return p1, p2}]"}
bhe-api-1       | {"level":"info","user_id":"7f63d6b4-c472-43f2-9ed1-5fac1a0fa6e8","remote_addr":"172.31.0.4:58282","proto":"HTTP/1.1","referer":"http://bhe.localhost/ui/mockServiceWorker.js","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36","request_id":"ebd37231-dab6-4043-bd20-4ff77fa293a9","request_bytes":266,"response_bytes":465,"status":500,"elapsed":1374.180417,"time":"2024-03-26T18:16:22.206038092Z","message":"POST /api/v2/graphs/cypher"}
```

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [ ] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

**Comment by juggernot325** [comment]
There's a bug in `IsNeoTimeoutError` that is causing it to not properly return true when the timeout error occurs at the driver level. I'm going to include a fix for that in this PR.

File: packages/go/dawgs/util/errors.go [hunk 1]
```diff
@@ -58,9 +60,14 @@ func (s *errorCollector) Combined() error {
 }
 
 func IsNeoTimeoutError(err error) bool {
-	if castError, ok := err.(*neo4j.Neo4jError); !ok {
+	switch e := err.(type) {
```
Comment (zinic) [inline packages/go/dawgs/util/errors.go:63]: Quick question here, does the `dawgs` package wrap the original Neo4j error or does it output its own? If it's wrapped we might be able to use the `errors` package to unwrap and get to the original error to tighten this logic up.

If not then this is best effort but it also may be worth seeing what the change looks like to get `dawgs` to wrap the originating error.

Comment (juggernot325) [reply packages/go/dawgs/util/errors.go:63]: Oh good call, I hadn't considered that. Let me play around with it locally and I'll clean it up if possible.

Comment (juggernot325) [reply packages/go/dawgs/util/errors.go:63]: I did some research here and I don't believe that wrapping would be an easy to implement solution here. The error is returned directly from Neo4j and put into a `graph.Error` which is a struct containing the query performed and the raw driver error. The underlying `neo4j-go-driver` library wraps some errors, but timeouts are not one of them.

I think I'll leave this as-is for now and we can look at a better system in a future PR.

**Review by elikmiller — APPROVED** [review]
Looks good to me!

**Pull request merged**